charms/reactive/bus.py

```diff
@@ -342,6 +344,13 @@ def _invoke(to_invoke):
                     break
         FlagWatch.commit()
 
+    # When in restricted context, only run hooks for that context.
+    if restricted:
+        unitdata.kv().set('reactive.dispatch.phase', 'restricted')
+        hook_handlers = _test(Handler.get_handlers())
+        _invoke(hook_handlers)
+        return
```

Comment:
An alternative to creating a new phase would be to just reuse the existing hook phase. A new phase is likely cleaner, but might derail some of the inflight reactive 2.0 work.